File: plugins/plugin-auto-trader/python/elizaos_plugin_auto_trader/actions/check_portfolio.py
from __future__ import annotations

from dataclasses import dataclass

from elizaos_plugin_auto_trader.actions.common import ActionResult, Message
from elizaos_plugin_auto_trader.service import TradingService


@dataclass
class CheckPortfolioAction:
    @property
    def name(self) -> str:
        return "CHECK_PORTFOLIO"

    async def validate(self, _message: Message, _state: dict) -> bool:  # type: ignore[type-arg]
        return True

    async def handler(
        self,
        _message: Message,
        _state: dict,  # type: ignore[type-arg]
        service: TradingService | None = None,
    ) -> ActionResult:
        if service is None:
            return ActionResult(False, "TradingService is not available.", "missing_service")

        portfolio = await service.check_portfolio()
        lines = [
            f"Portfolio: ${portfolio.total_value:.2f} "
            f"| PnL: ${portfolio.pnl:.2f} ({portfolio.pnl_pct:+.2f}%)"
        ]

        if portfolio.holdings:
            for token, h in portfolio.holdings.items():
                lines.append(
                    f"  {token}: {h.amount:.4f} @ ${h.current_price:.4f} "
                    f"= ${h.value:.2f} (PnL: ${h.pnl:.2f})"
                )
        else:
            lines.append("  No open positions.")

        return ActionResult(True, "\n".join(lines), data=portfolio)

File: plugins/plugin-auto-trader/rust/src/actions/check_portfolio.rs
use async_trait::async_trait;
use serde_json::{json, Value};

use crate::service::TradingService;
use crate::{Action, ActionExample, ActionResult};

pub struct CheckPortfolioAction;

#[async_trait]
impl Action for CheckPortfolioAction {
    fn name(&self) -> &str {
        "CHECK_PORTFOLIO"
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "VIEW_PORTFOLIO",
            "SHOW_PORTFOLIO",
            "GET_PORTFOLIO",
            "PORTFOLIO_STATUS",
        ]
    }

    fn description(&self) -> &str {
        "Check the current portfolio holdings and PnL."
    }

    async fn validate(&self, _message: &Value, _state: &Value) -> bool {
        true
    }

    async fn handler(
        &self,
        _message: &Value,
        _state: &Value,
        service: Option<&TradingService>,
    ) -> ActionResult {
        let Some(svc) = service else {
            return ActionResult {
                success: false,
                text: "TradingService is not available.".to_string(),
                data: None,
                error: Some("missing_service".to_string()),
            };
        };

        let portfolio = svc.check_portfolio().await;

        let mut lines = vec![format!(
            "Portfolio: ${:.2} | PnL: ${:.2} ({:+.2}%)",
            portfolio.total_value, portfolio.pnl, portfolio.pnl_pct
        )];

        for (token, h) in &portfolio.holdings {
            lines.push(format!(
                "  {}: {:.4} @ ${:.4} = ${:.2} (PnL: ${:.2})",
                token, h.amount, h.current_price, h.value, h.pnl
            ));
        }

        if portfolio.holdings.is_empty() {
            lines.push("  No open positions.".to_string());
        }

        ActionResult {
            success: true,
            text: lines.join("\n"),
            data: Some(json!(portfolio)),
            error: None,
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![ActionExample {
            user_message: "Check my portfolio".to_string(),
            agent_response: "Portfolio: $10,000.00 | PnL: $0.00 (+0.00%)".to_string(),
        }]
    }
}
